src/UserSpaceInstrumentation/AccessTraceesMemoryTest.cpp

```diff
@@ -32,33 +125,88 @@ TEST(AccessTraceesMemoryTest, ReadWriteRestore) {
   }
 
   // Stop the child process using our tooling.
-  CHECK(AttachAndStopProcess(pid).has_value());
+  CHECK(!AttachAndStopProcess(pid).has_error());
+
+  auto continuous_range_or_error = GetFirstContinuousAddressRange(pid);
+  CHECK(continuous_range_or_error.has_value());
+  const auto continuous_range = continuous_range_or_error.value();
+  const uint64_t address = continuous_range.first;
+  const uint64_t length = continuous_range.second - continuous_range.first;
+
+  // Backup.
+  auto backup = ReadTraceesMemory(pid, address, length);
+  CHECK(backup.has_value());
+
+  // Good write.
+  std::vector<uint8_t> bytes(length, 0);
+  ErrorMessageOr<void> result = WriteTraceesMemory(pid, address, bytes);
+  EXPECT_FALSE(result.has_error());
+
+  // Process does not exist.
+  result = WriteTraceesMemory(-1, address, bytes);
+  ASSERT_TRUE(result.has_error());
+  EXPECT_TRUE(absl::StrContains(result.error().message(), "Unable to open file"))
+      << result.error().message();
+
+  // Write 0 bytes.
+  EXPECT_DEATH(auto unused_result = WriteTraceesMemory(pid, address, std::vector<uint8_t>()),
+               "Check failed");
+
+  // Write past the end of the mappings.
+  bytes.push_back(0);
+  result = WriteTraceesMemory(pid, address, bytes);
+  ASSERT_TRUE(result.has_error());
+  EXPECT_TRUE(absl::StrContains(result.error().message(), "Input/output error"))
+      << result.error().message();
+
+  // Write to bad address.
+  result = WriteTraceesMemory(pid, 0, bytes);
+  ASSERT_TRUE(result.has_error());
+  EXPECT_TRUE(absl::StrContains(result.error().message(), "Input/output error"))
+      << result.error().message();
+
+  // Restore, detach and end child.
+  CHECK(!WriteTraceesMemory(pid, address, backup.value()).has_error());
+  CHECK(!DetachAndContinueProcess(pid).has_error());
+  kill(pid, SIGKILL);
+  waitpid(pid, NULL, 0);
+}
+
+TEST(AccessTraceesMemoryTest, ReadWriteRestore) {
+  pid_t pid = fork();
+  ASSERT_TRUE(pid != -1);
+  if (pid == 0) {
+    // Child just runs an endless loop.
+    while (true) {
+    }
+  }
+
+  // Stop the child process using our tooling.
+  ASSERT_TRUE(AttachAndStopProcess(pid).has_value());
 
-  auto result_memory_region = GetFirstExecutableMemoryRegion(pid);
-  CHECK(result_memory_region.has_value());
-  const uint64_t address_start = result_memory_region.value().first;
+  auto memory_region_or_error = GetFirstExecutableMemoryRegion(pid);
+  ASSERT_TRUE(memory_region_or_error.has_value());
```

Comment:
if it fails here - we loose the error message. Maybe we want to introduce a macro that checks `ErrorMessageOr<>`  and always adds the actual error message.

Reply:
I think the right way to get this into gtest would be to write a matchers to enable something like this:
```
EXPECT_THAT(result, ::testing::HasValue())
EXPECT_THAT(result, ::testing::HasError())
ASSERT_THAT(result, ::testing::HasErrorContaining("permission"))
```
I dislike the `<< result.error().message()` I added for its unneccessary verbosity.

Not in this PR thought: http://b/185464948